zilliqa/src/consensus.rs

```diff
@@ -712,7 +707,9 @@ impl Consensus {
     ) -> Result<Option<TransactionApplyResult>> {
         let hash = txn.hash;
 
-        self.db.insert_transaction(&hash, &txn.tx)?;
+        if !self.db.contains_transaction(&txn.hash)? {
+            self.db.insert_transaction(&txn.hash, &txn.tx)?;
```

Comment:
Are we ever, potentially, going to have the node run multi-threaded? Like a single consensus? So far everything we've built is geared towards a single thread per node. And if we're single-threaded, I'd favour explicit checks (and, where the check isn't present, the DB flagging an error if we try to double-insert) than just silently swallowing up every potential conflict at the DB level.

If we do ever multi-thread our nodes then this might need changing, of course. Would a comment be worthwhile? If multithreading isn't even in the plans then it might just be noise. Let me know if you'd like me to add one.